### PR TITLE
Component suggestions

### PR DESCRIPTION
**Pull request by adrian-ponce-wolox**
## Summary ##
**Trello card**: https://trello.com/c/cVFzfaNb/20-componente-de-sugeridos

Suggestions Component

## Test Plan ##

  * Open the app
  * Login with Google
  * The main tab bar view should appear after a successfull login.
  * Press on a Book cell
  * The Book Detail  view should appear with a Suggestions Cell
  * Book images should appear on the Suggestion Component

### iPhone 7
component_suggestions

## Screencast ##
component_suggestions

### CODE REVIEW

File: WBooks/Repositories/BooksRepository.swift [hunk 1]
```diff
@@ -43,12 +44,19 @@ final public class BooksRepository: BaseRepository {
             return decode(JSON).toResult()
         }
     }
+
+    func fetchSuggestions(bookID: Int) -> SignalProducer<[Book], RepositoryError> {
```
Comment (danielaRiesgo) [inline WBooks/Repositories/BooksRepository.swift]: `func fetchSuggestions(for bookID: Int) -> SignalProducer<[Book], RepositoryError>`

File: WBooks/Suggestions/SuggestionsTableViewCell.swift [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            switch result {
+            case .Success:
+                self.clearStack()
+                
+                let count: Int = min(SuggestionsTableViewCell.MaxBooksShown, viewModel.count)
+                for index: Int in 0 ..< count {
+                    self.addBookImageWith(viewModel[index].imageURL)
+                }
+            case .Failure:
+                print("Failed")
```
Comment (danielaRiesgo) [inline WBooks/Suggestions/SuggestionsTableViewCell.swift]: Better than printing, is logging with a good message ! That has a timestamp and all: `NSLog`.
If it couldn't bring any suggestions, should the cell still be shown? Should it have a text like `There is no suggestions for this book yet.`

File: WBooks/Suggestions/SuggestionsTableViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    @IBOutlet weak var stackView: UIStackView!
+    
+    func bindViewModel(viewModel: SuggestionsViewModel) {
+        viewModel.fetchSuggestions().startWithResult { [unowned self] result in
+            switch result {
+            case .Success:
+                self.clearStack()
+                
+                let count: Int = min(SuggestionsTableViewCell.MaxBooksShown, viewModel.count)
+                for index: Int in 0 ..< count {
```
Comment (danielaRiesgo) [inline WBooks/Suggestions/SuggestionsTableViewCell.swift]: The `: Int` is unnecessary.

File: WBooks/Suggestions/SuggestionsTableViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    @IBOutlet weak var titleLabel: UILabel!
+    @IBOutlet weak var stackView: UIStackView!
+    
+    func bindViewModel(viewModel: SuggestionsViewModel) {
+        viewModel.fetchSuggestions().startWithResult { [unowned self] result in
+            switch result {
+            case .Success:
+                self.clearStack()
+                
+                let count: Int = min(SuggestionsTableViewCell.MaxBooksShown, viewModel.count)
```
Comment (danielaRiesgo) [inline WBooks/Suggestions/SuggestionsTableViewCell.swift]: The `: Int` is unnecessary.

File: WBooks/Suggestions/SuggestionsTableViewCell.swift [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+private extension SuggestionsTableViewCell {
+
+    func clearStack() {
+        stackView.arrangedSubviews.forEach { $0.removeFromSuperview() }
+    }
+
+    func addBookImageWith(url: NSURL?) {
+        if let imageUrl = url {
+            let containerView = UIView(frame:CGRect())
```
Comment (danielaRiesgo) [inline WBooks/Suggestions/SuggestionsTableViewCell.swift]: Add space after `:`.

File: WBooks/Suggestions/SuggestionsTableViewCell.swift [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            stackView.addArrangedSubview(containerView)
+            
+            let imageView = UIImageView(frame: CGRect(x: 0, y: 0, width: 40, height: 60))
```
Comment (danielaRiesgo) [inline WBooks/Suggestions/SuggestionsTableViewCell.swift]: Use clear constants instead of "magic numbers".

Comment (danielaRiesgo) [inline WBooks/Suggestions/SuggestionsTableViewCell.swift]: Btw, this is unnecessary I think, if then you are constraining its width and height.

File: WBooks/Suggestions/SuggestionsViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+//
+
+import Foundation
+import ReactiveCocoa
+
+final class SuggestionsViewModel {
+    
+    private let _booksRepository: BooksRepository
+    
+    private var _books: [BookViewModel] = []
```
Comment (danielaRiesgo) [inline WBooks/Suggestions/SuggestionsViewModel.swift]: I would call it `_suggestions` or `_bookSuggestions` to not be confused with book where the suggestions come from.

File: WBooks/Suggestions/SuggestionsViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    private var _bookId: Int
+    
+    init(booksRepository: BooksRepository, bookId: Int) {
+        _booksRepository = booksRepository
+        _bookId = bookId
+    }
+    
+    func fetchSuggestions() -> SignalProducer<(), RepositoryError> {
```
Comment (danielaRiesgo) [inline WBooks/Suggestions/SuggestionsViewModel.swift]: If this was a `SignalProducer<[BookViewModel], RepositoryError>`, then I think storing the values in this class is not necessary. But I don't think it's bad if we store them either.

File: WBooks/Suggestions/SuggestionsViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        return _booksRepository.fetchSuggestions(_bookId)
+            .on(next: { [unowned self] suggestions in
+                self._books = suggestions.map { BookViewModel(book: $0) }
+                })
+            .map { _ in return }
```
Comment (danielaRiesgo) [inline WBooks/Suggestions/SuggestionsViewModel.swift]:
```
.on(next: { [unowned self] suggestions in
    self._books = suggestions.map { BookViewModel(book: $0) }
}).map { _ in return } 
```

**Comment by danielaRiesgo** [comment]
@adrian-ponce-wolox Something I've just noticed is that comment's cells have more height than they require. Search about `UITableViewAutomaticDimension` and add a new card to correct that.

File: WBooks/Suggestions/SuggestionsViewModel.swift (hunk 8 above)
Comment (adrian-ponce-wolox) [inline WBooks/Suggestions/SuggestionsViewModel.swift]: It's true. This would be more practical if it was the viewModel of a table or similar

Comment (adrian-ponce-wolox) [inline WBooks/Suggestions/SuggestionsViewModel.swift]: I have changed the table to show the cell only if there is at least one suggestion. So this takes relevance now

File: WBooks/Suggestions/SuggestionsViewModel.swift [hunk 10]
```diff
@@ -12,28 +12,26 @@ import ReactiveCocoa
...
     
     func fetchSuggestions() -> SignalProducer<(), RepositoryError> {
-        return _booksRepository.fetchSuggestions(_bookId)
+        return _booksRepository.fetchSuggestions(forBook: _bookId)
             .on(next: { [unowned self] suggestions in
-                self._books = suggestions.map { BookViewModel(book: $0) }
-                })
-            .map { _ in return }
+                self._suggestions = suggestions.map { BookViewModel(book: $0) }
+                }).map { _ in return }
```
Comment (danielaRiesgo) [inline WBooks/Suggestions/SuggestionsViewModel.swift]: Indent one left.

File: WBooks/BookDetail/BookDetailViewController.swift [hunk 11]
```diff
@@ -14,9 +14,10 @@ import enum Result.NoError
 
 enum SectionType: Int {
     case BookInfoSection
+    case SuggestionsSection
     case CommentSection
     public static var sectionsCount: Int {
```
Comment (dtilve) [inline WBooks/BookDetail/BookDetailViewController.swift]: Inline this as `public static var sectionsCount: Int = 3`

File: WBooks/BookDetail/BookDetailViewController.swift [hunk 12]
```diff
@@ -71,6 +74,16 @@ extension BookDetailViewController {
         }
     }
     
+    func fetchSuggestions() {
+        _viewModel.suggestionsViewModel.fetchSuggestions().startWithResult { [unowned self] result in
+            switch result {
+            case .Success: self._view.tableView.reloadData()
+            case .Failure:
+                self.showAlert("BOOK_DETAIL.FETCHING_SUGGESTIONS_ERROR".localized(), actionTitle: "ALERT.OK".localized())
```
Comment (dtilve) [inline WBooks/BookDetail/BookDetailViewController.swift]: Inline this.

File: WBooks/BookDetail/BookDetailViewController.swift [hunk 13]
```diff
@@ -105,6 +120,9 @@ extension BookDetailViewController: UITableViewDataSource {
         switch section {
         case SectionType.BookInfoSection.rawValue:
             return 1
+        case SectionType.SuggestionsSection.rawValue:
+            guard _viewModel.suggestionsViewModel.count > 0 else { return 0 }
```
Comment (dtilve) [inline WBooks/BookDetail/BookDetailViewController.swift]: return _viewModel.suggestionsViewModel.count > 0 ? 1 : 0

File: WBooks/SuggestBook/SuggestBookViewModel.h [hunk 14]
```diff
@@ -6,7 +6,7 @@
 //  Copyright © 2017 Wolox. All rights reserved.
 //
 
-@interface SuggestBookViewModel : NSObject 
+@interface SuggestBookViewModel: NSObject
```
Comment (dtilve) [inline WBooks/SuggestBook/SuggestBookViewModel.h]: Aren't you using `swiftlint`?

Comment (adrian-ponce-wolox) [inline WBooks/SuggestBook/SuggestBookViewModel.h]: Yes, but it doesn't work on objective-c files. :/

Comment (dtilve) [inline WBooks/SuggestBook/SuggestBookViewModel.h]: Duh makes sense

**Pull request merged**